apps/glusterfs/operations_block_volume.go

```diff
@@ -376,6 +376,288 @@ func (bvc *BlockVolumeCreateOperation) CleanDone() error {
+// expand an existing blockvolume.
+type blockVolExpandReclaim struct {
+	actualSize int
+}
+
+type BlockVolumeExpandOperation struct {
+	OperationManager
+	noRetriesOperation
+	bvolId string
+
+	// modification values
+	newSize int
+	reclaim blockVolExpandReclaim // gets set by Clean() call
+}
+
+// NewBlockVolumeExpandOperation creates a new BlockVolumeExpandOperation populated
+// with the given blockvolume id, db connection and newsize (in GB) that the
+// blockvolume is to be expanded by.
+func NewBlockVolumeExpandOperation(
+	bvolId string, db wdb.DB, newSize int) *BlockVolumeExpandOperation {
+
+	return &BlockVolumeExpandOperation{
+		OperationManager: OperationManager{
+			db: db,
+			op: NewPendingOperationEntry(NEW_ID),
+		},
+		bvolId:  bvolId,
+		newSize: newSize,
+	}
+}
+
+// loadBlockVolumeExpandOperation returns a BlockVolumeExpandOperation populated
+// from an existing pending operation entry in the db.
+func loadBlockVolumeExpandOperation(
+	db wdb.DB, p *PendingOperationEntry) (*BlockVolumeExpandOperation, error) {
+
+	bvolIds := []string{}
+	err := db.View(func(tx *bolt.Tx) error {
+		for _, a := range p.Actions {
+			switch a.Change {
+			case OpExpandBlockVolume:
+				bvolIds = append(bvolIds, a.Id)
+			}
+		}
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(bvolIds) != 1 {
+		return nil, fmt.Errorf(
+			"Incorrect number of block volume Ids (%v) for expand operation: %v",
+			len(bvolIds), p.Id)
+	}
+
+	return &BlockVolumeExpandOperation{
+		OperationManager: OperationManager{
+			db: db,
+			op: p,
+		},
+		bvolId: bvolIds[0],
+	}, nil
+}
+
+func (bve *BlockVolumeExpandOperation) Label() string {
+	return "Expand Block Volume"
+}
+
+func (bve *BlockVolumeExpandOperation) ResourceUrl() string {
+	return fmt.Sprintf("/blockvolumes/%v", bve.bvolId)
+}
+
+func (bve *BlockVolumeExpandOperation) Build() error {
+	var bv *BlockVolumeEntry
+	return bve.db.Update(func(tx *bolt.Tx) error {
+		var err error
+		bv, err = NewBlockVolumeEntryFromId(tx, bve.bvolId)
+		if err != nil {
+			return err
+		}
+		if bv.Pending.Id != "" {
+			logger.LogError("Pending block volume %v can not be Expanded",
+				bve.bvolId)
+			return ErrConflict
+		}
+
+		bhv, err := NewVolumeEntryFromId(tx, bv.Info.BlockHostingVolume)
+		if err != nil {
+			return err
+		}
+		if bhv.Pending.Id != "" {
+			logger.LogError("Can not expand block volume %v when hosting volume %v is pending",
+				bve.bvolId, bhv.Info.Id)
+			return ErrConflict
+		}
+
+		if bve.newSize == bv.Info.Size {
+			err := logger.LogError("Requested new-size %v is same as current "+
+				"block volume size %v, nothing to be done.", bve.newSize, bv.Info.Size)
+			return err
+		} else if bve.newSize < bv.Info.Size {
+			err := logger.LogError("Requested new-size %v is less than current "+
+				"block volume size %v, shrinking is not allowed.", bve.newSize, bv.Info.Size)
+			return err
+		}
+
+		requiredFeeSize := bve.newSize - bv.Info.Size
```

Comment:
Small typo but it also gives a new meaning to the statement.

Reply:
ok